framework/src/parser/Parser.C

```diff
@@ -139,425 +238,320 @@ Parser::getFileName(bool stripLeadingPath) const
-  std::set_difference(difference.begin(),
-                      difference.end(),
-                      requested_vars.begin(),
-                      requested_vars.end(),
-                      std::inserter(no_overrides, no_overrides.end()));
-
-  if (!no_overrides.empty())
+  try
   {
-    std::ostringstream oss;
+    _root = hit::parse(_input_filename, input);
 
-    oss << "The following parameters were unused "
-        << (in_input_file ? "in your input file:\n" : "on the command line:\n");
-    for (const auto & name : no_overrides)
-      oss << name << "\n";
+    int argc = _app.commandLine()->argc();
+    char ** argv = _app.commandLine()->argv();
+    auto cli_input = hitCLIFilter(_app.name(), argc, argv);
 
-    if (error_on_warn)
-      mooseError(oss.str());
-    else
-      mooseWarning(oss.str());
+    _cli_root = hit::parse("CLI_ARGS", cli_input);
+    hit::explode(_cli_root);
+    hit::explode(_root);
+    hit::merge(_cli_root, _root);
   }
-}
-
-void
-Parser::checkOverriddenParams(bool error_on_warn) const
-{
-  if (!_sections_read && error_on_warn)
-    // The user has requested errors but we haven't done any parsing yet so throw an error
-    mooseError("No parsing has been done, so checking for overridden parameters is not possible");
-
-  std::set<std::string> overridden_vars = _getpot_file_error_checking.get_overridden_variables();
-
-  if (!overridden_vars.empty())
+  catch (hit::ParseError & err)
   {
-    std::ostringstream oss;
-
-    oss << "The following variables were overridden or supplied multiple times:\n";
-    for (const auto & name : overridden_vars)
-      oss << name << "\n";
+    mooseError(err.what());
+  }
 
-    if (error_on_warn)
-      mooseError(oss.str());
-    else
-      mooseWarning(oss.str());
+  // expand ${bla} parameter values and mark/include variables used in expansions as "used".  This
+  // MUST occur before parameter extraction - otherwise parameters will get wrong values.
+  ExpandWalker exw(_input_filename);
+  _root->walk(&exw);
+  for (auto & var : exw.used)
+    _extracted_vars.insert(var);
+  for (auto & msg : exw.errors)
+    _errmsg += msg + "\n";
+
+  // do as much error checking as early as possible so that errors are more useful instead
+  // of surprising and disconnected from what caused them.
+  DupParamWalker dw(_input_filename);
+  BadActiveWalker bw(_input_filename);
+  _root->walk(&dw, hit::NodeType::Field);
+  _root->walk(&bw, hit::NodeType::Section);
+  for (auto & msg : dw.errors)
+    _errmsg += msg + "\n";
+  for (auto & msg : bw.errors)
+    _errmsg += msg + "\n";
+
+  if (_errmsg.size() > 0)
+    mooseError(_errmsg);
+
+  // There are a few order dependent actions that have to be built first in
+  // order for the parser and application to function properly:
+  //
+  // SetupDebugAction: This action can contain an option for monitoring the parser progress. It must
+  // be parsed first
+  //                   to capture all of the parsing output.
```

Comment:
Flow got kind of messed up here when we changed line length. Perhaps you can fix it?